eclair-core/src/main/scala/fr/acinq/eclair/payment/PaymentPacket.scala

```diff
@@ -42,40 +42,68 @@ object IncomingPaymentPacket {
     Sphinx.peel(privateKey, Some(paymentHash), packet) match {
-      case Right(p@Sphinx.DecryptedPacket(payload, nextPacket, _)) =>
-        perHopPayloadCodec(p.isLastPacket).decode(payload.bits) match {
-          case Attempt.Successful(DecodeResult(perHopPayload, _)) => Right(DecodedOnionPacket(perHopPayload, nextPacket))
-          case Attempt.Failure(e: OnionRoutingCodecs.MissingRequiredTlv) => Left(e.failureMessage)
-          case Attempt.Failure(e: OnionRoutingCodecs.ForbiddenTlv) => Left(e.failureMessage)
-          // Onion is correctly encrypted but the content of the per-hop payload couldn't be parsed.
-          // It's hard to provide tag and offset information from scodec failures, so we currently don't do it.
-          case Attempt.Failure(_) => Left(InvalidOnionPayload(UInt64(0), 0))
+      case Right(p: Sphinx.DecryptedPacket) =>
+        PaymentOnionCodecs.perHopPayloadCodec.decode(p.payload.bits) match {
+          case Attempt.Successful(DecodeResult(perHopPayload, _)) if p.isLastPacket => Right(DecodedOnionPacket(perHopPayload, None))
+          case Attempt.Successful(DecodeResult(perHopPayload, _)) => Right(DecodedOnionPacket(perHopPayload, Some(p.nextPacket)))
+          case Attempt.Failure(_) =>
+            // Onion is correctly encrypted but the content of the per-hop payload couldn't be decoded.
+            // It's hard to provide tag and offset information from scodec failures, so we currently don't do it.
+            Left(InvalidOnionPayload(UInt64(0), 0))
         }
       case Left(badOnion) => Left(badOnion)
     }
 
+  case class DecodedEncryptedRecipientData(payload: TlvStream[RouteBlindingEncryptedDataTlv], nextBlinding: PublicKey)
+
+  private[payment] def decryptEncryptedRecipientData(add: UpdateAddHtlc, privateKey: PrivateKey, payload: TlvStream[OnionPaymentPayloadTlv], encryptedRecipientData: ByteVector): Either[FailureMessage, DecodedEncryptedRecipientData] = {
+    if (add.blinding_opt.isDefined && payload.get[OnionPaymentPayloadTlv.BlindingPoint].isDefined) {
+      // TODO: return an unparseable error
+      Left(InvalidOnionPayload(UInt64(12), 0))
+    } else {
+      add.blinding_opt.orElse(payload.get[OnionPaymentPayloadTlv.BlindingPoint].map(_.publicKey)) match {
+        case Some(blinding) => RouteBlindingEncryptedDataCodecs.decode(privateKey, blinding, encryptedRecipientData) match {
+          case Left(_) =>
+            // There are two possibilities in this case:
+            //  - the blinding point is invalid: the sender or the previous node is buggy or malicious
+            //  - the encrypted data is invalid: the sender, the previous node or the recipient must be buggy or malicious
+            // TODO: return an unparseable error
+            Left(InvalidOnionPayload(UInt64(12), 0))
+          case Right(decoded) => Right(DecodedEncryptedRecipientData(decoded.tlvs, decoded.nextBlinding))
+        }
+        case None =>
```

Comment:
I would have handled this in the `decrypt` method and simplify the signature of `decryptEncryptedRecipientData`

Reply:
I'm not convinced, this is really a validation step that is tied to the encrypted data, I think it makes sense to have it in this function...also if done in `decrypt`, we'd need to duplicate that check (since we do it for the intermediate and final case) or re-order them, which I don't like (I think the separation should be intermediate vs final *before* separating standard vs blinded, even though it's arguable).